docarray/array/array.py

```diff
@@ -13,34 +38,326 @@ class DocumentArray(
+
+    You should use stacked mode, if you:
+
+    * You want to process the entire DocumentArray as one batch
+    * Are using the DocumentArray in an ML model for training or inference
+
     """
 
     document_type: Type[BaseDocument] = AnyDocument
 
     def __init__(self, docs: Iterable[BaseDocument]):
         super().__init__(doc_ for doc_ in docs)
 
+        self._columns: Optional[
+            Dict[str, Union[TorchTensor, AbstractDocumentArray, NdArray, None]]
+        ] = None
+
     def __class_getitem__(cls, item: Type[BaseDocument]):
         if not issubclass(item, BaseDocument):
             raise ValueError(
                 f'DocumentArray[item] item should be a Document not a {item} '
             )
 
-        class _DocumenArrayTyped(DocumentArray):
+        class _DocumentArrayTyped(cls):  # type: ignore
             document_type: Type[BaseDocument] = item
 
-        for field in _DocumenArrayTyped.document_type.__fields__.keys():
+        for field in _DocumentArrayTyped.document_type.__fields__.keys():
 
             def _property_generator(val: str):
-                return property(lambda self: self._get_documents_attribute(val))
+                def _getter(self):
+                    return self._get_array_attribute(val)
+
+                def _setter(self, value):
+                    self._set_array_attribute(val, value)
 
-            setattr(_DocumenArrayTyped, field, _property_generator(field))
+                # need docstring for the property
+                return property(fget=_getter, fset=_setter)
+
+            setattr(_DocumentArrayTyped, field, _property_generator(field))
             # this generates property on the fly based on the schema of the item
 
-        _DocumenArrayTyped.__name__ = f'DocumentArray[{item.__name__}]'
-        _DocumenArrayTyped.__qualname__ = f'DocumentArray[{item.__name__}]'
+        _DocumentArrayTyped.__name__ = f'DocumentArray[{item.__name__}]'
+        _DocumentArrayTyped.__qualname__ = f'DocumentArray[{item.__name__}]'
+
+        return _DocumentArrayTyped
+
+    def __getitem__(self, item):
+        if self.is_stacked():
+            doc = super().__getitem__(item)
+            # NOTE: this could be speed up by using a cache
+            for field in self._columns.keys():
+                setattr(doc, field, self._columns[field][item])
+            return doc
+        else:
+            return super().__getitem__(item)
+
+    def stack(self: T) -> T:
+        """
+        Puts the DocumentArray into stacked mode.
+        :return: itself
+
+        When entering stacked mode the DocumentArray creates a column for each field
+        of the Document that are Tensors or nested Documents that contain at
+        least one Tensor field. This is useful when to perform operations on the whole
+        array at once. In stacked mode, accessing or setting the DocumentArray's fields
+        DocumentArray accesses or sets the column of the array.
```

Comment:
```suggestion
        When entering stacked mode DocumentArray creates a column for:
        
         - Each Tensor field of each Document
         - Nested Documents that contain at least one Tensor field. 
         
        This is useful to perform operations on the whole
        array at once. In stacked mode, accessing or setting the DocumentArray's fields
        accesses or sets the columns of the array.
```

Can you confirm this is correct interpretation? There was some odd wording to fix.